---
title: "Basics of {rrrSingleCellUtils}"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Basics of rrrSingleCellUtils}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(rrrSingleCellUtils)
```

## Processing a Seurat object
Run the standard processing steps in a single command
```{r eval=TRUE}
pbmc_fixed <-
    process_seurat(SeuratObject::pbmc_small,
                   run_umap_dims = 1:10,
                   resolution = 0.8)
Seurat::DimPlot(pbmc_fixed)
```

## Using SNPs to define tumor clusters
```{r eval=FALSE}
two_sobj <- qs::qread("/home/gdrobertslab/mvc002/analyses/roberts/dev/testSnps/output/patient_met_1/two_sobj.qs")

c_b_t <-
    two_sobj@meta.data %>%
    select(used_clusters) %>%
    rename(cell_group = used_clusters) %>%
    rownames_to_column("cell_barcode") %>%
    as_tibble() %>%
    mutate(bam_file = paste0("/home/gdrobertslab/lab/Counts/",
                        str_remove(cell_barcode, "_.*"),
                        "/outs/possorted_genome_bam.bam"),
            cell_barcode = str_remove(cell_barcode, ".+_"))

get_snp_tree(cellid_bam_table = c_b_t,
             ploidy = "GRCh37",
             ref_fasta = "/home/gdrobertslab/lab/GenRef/10x-human/fasta/genome.fa",
             min_depth = 5,
             submit = TRUE)

```